src/Symfony/Component/Messenger/Bridge/AmqpExt/Middleware/RejectRedeliveredMessageMiddleware.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Bridge\AmqpExt\Middleware;
+
+use Symfony\Component\Messenger\Bridge\AmqpExt\Transport\AmqpReceivedStamp;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\RejectRedeliveredMessageException;
+use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+use Symfony\Component\Messenger\Middleware\StackInterface;
+use Symfony\Component\Messenger\Transport\AmqpExt\AmqpReceivedStamp as LegacyAmqpReceivedStamp;
+
+/**
+ * Middleware that throws a RejectRedeliveredMessageException when a message is detected that has been redelivered by AMQP.
+ *
+ * The middleware runs before the HandleMessageMiddleware and prevents redelivered messages from being handled directly.
+ * The thrown exception is caught by the worker and will trigger the retry logic according to the retry strategy.
+ *
+ * AMQP redelivers messages when they do not get acknowledged or rejected. This can happen when the connection times out
+ * or an exception is thrown before acknowledging or rejecting. When such errors happen again while handling the
+ * redelivered message, the message would get redelivered again and again. The purpose of this middleware is to prevent
+ * infinite redelivery loops and to unblock the queue by republishing the redelivered messages as retries with a retry
+ * limit and potential delay.
+ *
+ * @author Tobias Schultze <http://tobion.de>
+ */
+class RejectRedeliveredMessageMiddleware implements MiddlewareInterface
+{
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $amqpReceivedStamp = $envelope->last(AmqpReceivedStamp::class);
+        if ($amqpReceivedStamp instanceof AmqpReceivedStamp && $amqpReceivedStamp->getAmqpEnvelope()->isRedelivery()) {
+            throw new RejectRedeliveredMessageException('Redelivered message from AMQP detected that will be rejected and trigger the retry logic.');
```

Comment:
The original idea was that the RejectRedeliveredMessageMiddleware can be implemented also for other transports like doctrine. So not necessarily specific to amqp. But I guess we can then also just add a new middleware in in the doctrine bridge if needed at some point.
So it's fine to move this to the bridge. But then I would also suggest to move the `RejectRedeliveredMessageException` to the amqp bridge.

Reply:
Another thing is that we would need to conditionally enable this middleware when the bridge is installed. Currently it's always active but in 6.0 it would just error when the amqp bridge is not installed.

Reply:
I moved back the middleware to the core package. I did it for the following reasons: 
1) It was the original idea to let other transports use this middleware
2) I cannot see an easy way to conditionally add this as default middleware base on if it is installed or not.

Someone may decide to move it to the bridge in the future. But Im leaving it here for now.

Reply:
> I cannot see an easy way to conditionally add this as default middleware base on if it is installed or not.

Isn't that as simple as wrapping https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1584 around a class_exists call?

Reply:
There are two classes you need to check. You cannot simply wrap it, you need to write something like this: 

```php
$before[] = ['id' => 'add_bus_name_stamp_middleware'];
if (class_exists(LegacyMiddleware) || class_exists(BridgeyMiddleware)) {
    $before[] = ['id' => 'reject_redelivered_message_middleware'];
}
$before[] = ['id' => 'dispatch_after_current_bus'];
$before[] = ['id' => 'failed_message_processing_middleware'];

$defaultMiddleware = [
    'before' => $before,
    'after' => [
        ['id' => 'send_message'],
        ['id' => 'handle_message'],
    ],
];
```

I didnt feel like that was a very pretty solution (Maybe the "prettiest", but not "pretty"). So I decided to not move it, since you said that the idea was to let other transports add their logic here.